### DISCUSSION TITLE
Mixing async and sync code

### DISCUSSION THREAD

**Discussion by alexander-sony** [agronholm/anyio#584] [Q&A] [answered]
What is the recommended way of mixing async and sync code using anyio? See the discussion in https://bbc.github.io/cloudfit-public-docs/asyncio/asyncio-part-5.html, where `asyncio.get_event_loop().run_in_executor(send_request)` is used to handle a blocking call from a library that does not support async/await.

Thank's
Alex

**Comment by agronholm** [comment]
Have you read [AnyIO's documentation on the subject](https://anyio.readthedocs.io/en/3.x/threads.html)? In particular, `anyio.to_thread.run_sync()`?

**Reply by alexander-sony** [reply]
Would you have done something like this?

````
import time

import anyio

async def counter():
    now = time.time()
    print("Started counter")
    for i in range(0, 10):
        last = now
        await anyio.sleep(0.01)
        now = time.time()
        print(f"{i}: Was asleep for {now - last}s")

def blocker(id, t):
    print(f'blocker {id} to sleep for {t}s...')
    time.sleep(t)
    print(f'blocker {id} finished')

async def main():
    async with anyio.create_task_group() as tg:
        tg.start_soon(counter)
        await anyio.to_thread.run_sync(blocker, 1, 5)

anyio.run(main)

OUTPUT:
Started counter
blocker 1 to sleep for 5s...
0: Was asleep for 0.010510683059692383s
1: Was asleep for 0.01026463508605957s
2: Was asleep for 0.010253667831420898s
3: Was asleep for 0.010190725326538086s
4: Was asleep for 0.010202884674072266s
5: Was asleep for 0.010191202163696289s
6: Was asleep for 0.010184049606323242s
7: Was asleep for 0.010196208953857422s
8: Was asleep for 0.010170936584472656s
9: Was asleep for 0.010210275650024414s
blocker 1 finished
````

**Reply by agronholm** [reply, marked as answer]
Looks legit. What's the actual use case though?

**Reply by alexander-sony** [reply]
I have a concurrent application but it must also use some resources from a module that is not supporting async/await. Until now I have used threads, but would like to port it to use anyio/trio.

Thanks for your help!
